numpy/lib/function_base.py

```diff
@@ -1643,9 +1643,14 @@ def trim_zeros(filt, trim='fb'):
 
 def _trim_zeros_new(filt, trim='fb'):
     """Newer optimized implementation of ``trim_zeros()``."""
-    arr = np.asanyarray(filt).astype(bool, copy=False)
-
-    if arr.ndim != 1:
+    arr_any = np.asanyarray(filt)
+    arr = arr_any != 0 if arr_any.dtype != bool else arr_any
+
+    if arr is False:
```

Comment:
Certain data types do not support elementwise comparisons with.
Note that the intersection between aforementioned dtypes and those who cannot be cast into boolean arrays is pretty large. In this sense little changes with respect to the previous pull request: `str`, `bytes` & structured arrays still aren't supported.